Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift

```diff
@@ -0,0 +1,156 @@
+///
+/// This will continuously emit snapshots whenever the data changes.
+///
+/// ## Example Usage
+///
+/// ```
+/// // Create the snapshots provider
+/// let query = FetchResultSnapshotsProvider<StorageOrder>.Query(
+///     sortDescriptor: .init(keyPath: \StorageOrder.number, ascending: false)
+/// )
+/// let provider = FetchResultSnapshotsProvider(storageManager: storageManager, query: query)
+///
+/// // Set up the TableView and DataSource
+/// let dataSource =
+///    UITableViewDiffableDataSource<String, FetchResultSnapshotObjectID>(tableView: tableView,
+///                                                                       cellProvider: self.makeCellProvider())
+/// tableView.dataSource = dataSource
+///
+/// // Apply the emitted snapshots to the TableView's DataSource
+/// provider.snapshot.sink { snapshot in
+///     dataSource.apply(snapshot)
+/// }.store(in: &self.cancellables)
+///
+/// provider.start()
+/// ```
+///
+/// That's it! The `UITableView` should be automatically updated whenever the data changes.
+///
+@available(iOS 13.0, *)
+public final class FetchResultSnapshotsProvider<MutableType: FetchResultSnapshotsProviderMutableType>: NSObject, NSFetchedResultsControllerDelegate {
+
+    /// Defines the conditions for fetching the results.
+    public struct Query {
+        /// Defines how to sort the results.
+        ///
+        /// This is required because `NSFetchedResultsController` requires it.
+        ///
+        /// This needs to be extended to allow an array. However, we have to add protection that
+        /// there will always be at least one `NSSortDescriptor`. It'd be great if we can check
+        /// this requirement during compile-time.
+        public let sortDescriptor: NSSortDescriptor
+        /// An optional predicate to filter the results.
+        public let predicate: NSPredicate?
+        /// A keypath that returns the value to use for grouping results into sections.
+        ///
+        /// The values will be accessible in `FetchResultSnapshot.sectionIdentifiers`. In the UI,
+        /// those values can be converted to a user-friendly value.
+        public let sectionNameKeyPath: String?
+
+        init(sortDescriptor: NSSortDescriptor, predicate: NSPredicate? = nil, sectionNameKeyPath: String? = nil) {
+            self.sortDescriptor = sortDescriptor
+            self.predicate = predicate
+            self.sectionNameKeyPath = sectionNameKeyPath
+        }
+    }
+
+    /// The `StorageType` to perform the fetch on.
+    ///
+    /// This is currently forced to be the `viewStorage` (main thread). In the future, we can
+    /// probably move this to a background `StorageType` since `UITableViewDiffableDataSource`
+    /// allows consuming snapshots in the background.
+    private let storage: StorageType
+    /// The conditions to use when fetching the results.
+    ///
+    /// In the future, we can allow this to be mutable if necessary.
```

Comment:
yea, I think the query can change where there is sorting/filtering UI like the products tab. I also see the invalid number of rows crash in `ProductsViewController` (by searching ProductsViewController in Sentry) so it'd be nice if this `FetchResultSnapshotsProvider` could work for the products tab as well. Feel free to add the mutable support separately or whenever we get to add DiffableDataSource to the products tab!